universe/Condition.cpp

```diff
@@ -9956,6 +9965,249 @@ unsigned int Not::GetCheckSum() const {
+    if (this == &rhs)
+        return true;
+    if (typeid(*this) != typeid(rhs))
+        return false;
+
+    const OrderedAlternativesOf& rhs_ = static_cast<const OrderedAlternativesOf&>(rhs);
+
+    if (m_operands.size() != rhs_.m_operands.size())
+        return false;
+    for (unsigned int i = 0; i < m_operands.size(); ++i) {
+        CHECK_COND_VREF_MEMBER(m_operands.at(i))
+    }
+
+    return true;
+}
+
+void OrderedAlternativesOf::Eval(const ScriptingContext& parent_context,
+                                 ObjectSet& matches, ObjectSet& non_matches,
+                                 SearchDomain search_domain/* = NON_MATCHES*/) const
+{
+    std::shared_ptr<const UniverseObject> no_object;
+    ScriptingContext local_context(parent_context, no_object);
+
+    if (m_operands.empty()) {
+        ErrorLogger() << "OrderedAlternativesOf::Eval given no operands!";
+        return;
+    }
+    for (auto& operand : m_operands) {
+        if (!operand) {
+            ErrorLogger() << "OrderedAlternativesOf::Eval given null operand!";
+            return;
+        }
+    }
+
+    // OrderedAlternativesOf [ A B C ] matches all candidates which match the topmost condition.
+    // If any candidate matches A, then all candidates that match A are matched,
+    // or if no candidate matches A but any candidate matches B, then all candidates that match B are matched,
+    // or if no candidate matches A or B but any candidate matches C, then all candidates that match C are matched.
+    // If no candidate matches A, B, or C, then nothing is matched.
+    //
+    // Not OrderedAlternativesOf [ A B C ] finds the topmost condition which has matches and then matches its non-matches.
+    // If any candidate matches A, then all candidates that do not match A are matched,
+    // or if no candidates match A but any candidate matches B, then all candidates that do not match B are matched,
+    // or if no candidate matches A or B but any candidate matches C, then all candidates that do not match C are matched.
+    // If no candidate matches A, B, or C, then all candidates are matched.
+    if (search_domain == NON_MATCHES) {
+        // check all operand conditions on all objects in the non matches set
+        // stop if the operand condition has at least one match
+        // Return non-matches if there are any
+        ObjectSet temp_non_matches;
+        temp_non_matches.reserve(non_matches.size());
+        // try matching operand conditions until one matches
+        bool found_sth = false;
+        for (auto& operand : m_operands) {
+            // move those non matching the current operand to temp_non_matches
+            //            operand->Eval(local_context, matches, temp_non_matches, MATCHES);
+            // move those matching the current non-negated operand from non_matches to temp_non_matches
+            operand->Eval(local_context, non_matches, temp_non_matches, MATCHES);
```

Comment:
could probably search with `NON_MATCHES` here, with `non_matches` as the 3rd param and a temp container as the 2nd. Then if anything is in the temp container, can just move them into `matches` and don't need to touch `non_matches` again.

Reply:
Hm. I think my code is more complex than the problem. I would like to refactor to something more simple. I am halfway through thinking, please correct my reasoning:
the search domains are there for execution optimization (so you usually do not have to e.g. check an object for a match if you already know it fails to match a sibling in the surrounding AND condition. And you do not need to check an object for a non-match if you know it matches a sibling in the surrounding OR condition).  
 For the OrderedAlternativesOf in worst case I have to check all in the matches and non_matches set for a match, to find out if a operand condition is the one to determine matches.

Reply:
I thought that was what it does already, except checking both `matches` and `non_matches` when deciding which subcondition to use...